src/lib/ast.rs

```diff
@@ -0,0 +1,617 @@
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+#![warn(missing_docs)]
+
+use std::borrow::Cow;
+use std::collections::HashMap;
+use std::convert::{TryFrom, TryInto};
+use std::fmt;
+use std::fs::File;
+use std::io;
+use std::io::{Read, Write};
+use std::ops::{Index, IndexMut};
+use std::path::Path;
+
+use dot;
+use lrlex::{build_lex, Lexeme};
+use lrpar::parser;
+use lrtable::{Grammar, Minimiser, TIdx, yacc_to_statetable};
+
+#[derive(Debug)]
+/// Errors raised when parsing a source file.
+pub enum ParseError {
+    /// Io error returned from standard library routine.
+    Io(io::Error),
+    /// File not found.
+    FileNotFound,
+    /// File name had no extension (used to determine which lexer/parser to use).
+    NoFileExtension,
+    /// Lexer could not be built by `lrlex`.
+    BrokenLexer,
+    /// Parser could not be built by `lrpar`.
+    BrokenParser,
+    /// File contained lexical error and could not be lexed.
+    LexicalError,
+    /// File contained syntax error and could not be parsed.
+    SyntaxError,
+}
+
+/// A node identifier for nodes within a particular `Arena`.
+type NodeId = usize;
+
+/// A type to represent edges between nodes within an `Arena`.
+type EdgeId = (usize, usize);
+
+/// An arena in which to place `NodeID`-indexed AST nodes.
+///
+/// `T` and `U` are types given to internal values stored within each node,
+/// these information that the `lrpar` crate will provide, and represent the
+/// type (e.g. `expr`, `factor`, `term`) and information (e.g. literals) stored
+/// within terminals and non-terminals in a given grammar. In most uses, `T` and
+/// `U` will both be `String` types.
+pub struct Arena<T, U> {
+    nodes: Vec<Node<T, U>>,
+}
+
+impl<T, U> Index<NodeId> for Arena<T, U> {
+    type Output = Node<T, U>;
+
+    fn index(&self, node: NodeId) -> &Node<T, U> {
+        &self.nodes[node]
+    }
+}
+
+impl<T, U> IndexMut<NodeId> for Arena<T, U> {
+    fn index_mut(&mut self, node: NodeId) -> &mut Node<T, U> {
+        &mut self.nodes[node]
+    }
+}
+
+impl<T, U> Arena<T, U> {
+    /// Create an empty `Arena`.
+    pub fn new() -> Arena<T, U> {
+        Arena { nodes: Vec::new() }
+    }
+
+    /// Create a new node from its data.
+    pub fn new_node(&mut self, data: T, ty: U, indent: u16) -> NodeId {
+        let next_index = self.nodes.len();
+        let node = Node::new(data, ty, indent);
+        self.nodes.push(node);
+        next_index
+    }
+
+    /// Return `true` if the arena is empty, `false` otherwise.
+    pub fn is_empty(&self) -> bool {
+        self.nodes.is_empty()
+    }
+
+    /// Make one node the child of another.
+    /// Panic if either `NodeId` is not found.
```

Comment:
In general, the more typical Rust way is to say this returns something like `Error<()>`, so callers can decide whether they want to panic or not. [I'n not saying that's the right design or not here, but it may be worth thinking about.]